File: day6/solve.py
from math import prod
from re import findall, sub
from runner.python import solve


def part1(input: str) -> int:
    parsed = ((int(x) for x in findall(r"\d+", line)) for line in input.splitlines())
    part1 = prod(
        sum(1 for hold in range(0, time) if ((time - hold) * hold) > record)
        for time, record in zip(*parsed)
    )
    return part1


def part2(input: str) -> int:
    time, record = (int(sub(r"[^0-9]", "", line)) for line in input.splitlines())
    part2 = sum(1 for hold in range(0, time) if ((time - hold) * hold) > record)
    return part2


solve(
    part1,
    "Time:      7  15   30\n    Distance:  9  40  200",
    288,
    part2,
    "Time:      7  15   30\n    Distance:  9  40  200",
    71503,
)

File: day6/solve.ts
import { solve } from "../runner/typescript";
import { zip } from "lodash";

function parser(input: string) {
  return input
    .trim()
    .split("\n")
    .map((l) => l.match(/\d+/g).map(Number));
}

function getWins([time, distance]: [number, number]): number {
  const holds = [...Array(time).keys()];
  const distances = holds.map((hold) => (time - hold) * hold);
  const wins = distances.filter((d) => d > distance);
  return wins.length;
}

function part1([times, distances]: number[][]): number {
  return zip(times, distances)
    .map(getWins)
    .reduce((a, b) => a * b);
}

function part2([times, distances]: number[][]): number {
  const time = Number(times.join(""));
  const distance = Number(distances.join(""));
  return getWins([time, distance]);
}

solve({
  parser,
  part1,
  part2,
  part1Tests: [["Time:      7  15   30\n    Distance:  9  40  200\n    ", 288]],
  part2Tests: [["Time:      7  15   30\nDistance:  9  40  200", 71503]],
});
